File: packages/core/ios/HybridFFmpeg.swift
import Foundation
import NitroModules

class HybridFFmpeg: HybridFFmpegSpec {
    private let sessionQueue = OperationQueue()
    private let lock = NSLock()

    var onProgress: (_ sessionId: String, _ progress: NativeProgress) -> Void = { _, _ in }
    var onLog: (_ sessionId: String, _ log: NativeLogEntry) -> Void = { _, _ in }
    var onComplete: (_ sessionId: String, _ result: NativeSessionResult) -> Void = { _, _ in }

    override init() {
        super.init()
        sessionQueue.maxConcurrentOperationCount = 4
        sessionQueue.qualityOfService = .userInitiated
    }

    func execute(sessionId: String, args: [String], logLevel: Double) throws {
        let capturedOnProgress = onProgress
        let capturedOnLog = onLog
        let capturedOnComplete = onComplete

        sessionQueue.addOperation {
            let argc = Int32(args.count)
            let cArgs = args.map { strdup($0) }
            defer { cArgs.forEach { free($0) } }

            var argv = cArgs.map { UnsafePointer($0) }

            let context = CallbackContext(
                sessionId: sessionId,
                onProgress: capturedOnProgress,
                onLog: capturedOnLog
            )
            let contextPtr = Unmanaged.passRetained(context).toOpaque()

            var config = FFmpegConfig()
            config.session_id = sessionId.withCString { strdup($0) }
            config.on_progress = { sessionIdPtr, progress, userData in
                guard let userData = userData,
                      let sessionIdPtr = sessionIdPtr else { return }
                let ctx = Unmanaged<CallbackContext>.fromOpaque(userData).takeUnretainedValue()
                let sessionId = String(cString: sessionIdPtr)
                let nativeProgress = NativeProgress(
                    frame: Double(progress.frame),
                    fps: progress.fps,
                    bitrate: progress.bitrate,
                    totalSize: Double(progress.total_size),
                    timeMs: Double(progress.time_ms),
                    speed: progress.speed
                )
                ctx.onProgress(sessionId, nativeProgress)
            }
            config.on_log = { sessionIdPtr, logEntry, userData in
                guard let userData = userData,
                      let sessionIdPtr = sessionIdPtr else { return }
                let ctx = Unmanaged<CallbackContext>.fromOpaque(userData).takeUnretainedValue()
                let sessionId = String(cString: sessionIdPtr)
                let message = logEntry.message != nil ? String(cString: logEntry.message) : ""
                let nativeLog = NativeLogEntry(
                    level: Double(logEntry.level),
                    message: message
                )
                ctx.onLog(sessionId, nativeLog)
            }
            config.user_data = contextPtr
            config.log_level = Int32(logLevel)

            let result = ffmpeg_execute(argc, &argv, config)

            free(UnsafeMutablePointer(mutating: config.session_id))
            Unmanaged<CallbackContext>.fromOpaque(contextPtr).release()

            let logs = result.logs != nil ? String(cString: result.logs) : ""
            let failureMessage = result.failure_message != nil ? String(cString: result.failure_message) : nil

            var mutableResult = result
            ffmpeg_result_free(&mutableResult)

            let nativeResult = NativeSessionResult(
                sessionId: sessionId,
                returnCode: Double(result.return_code),
                duration: result.duration_ms,
                logs: logs,
                command: args,
                failureMessage: failureMessage
            )
            capturedOnComplete(sessionId, nativeResult)
        }
    }

    func cancel(sessionId: String) throws {
        sessionId.withCString { ptr in
            ffmpeg_cancel(ptr)
        }
    }

    func cancelAll() throws {
        ffmpeg_cancel_all()
    }

    func getActiveSessions() throws -> [String] {
        // Active sessions are tracked on the JS side;
        // native side manages cancellation state only
        return []
    }

    func probe(path: String) throws -> Promise<String> {
        return Promise.async {
            let result = path.withCString { ptr -> String in
                guard let jsonPtr = ffprobe_execute(ptr) else {
                    return "{\"error\": \"FFprobe returned null\"}"
                }
                let json = String(cString: jsonPtr)
                ffmpeg_string_free(jsonPtr)
                return json
            }
            return result
        }
    }

    func getFFmpegVersion() throws -> String {
        guard let versionPtr = ffmpeg_version() else {
            return "unknown"
        }
        return String(cString: versionPtr)
    }

    func getSupportedEncoders() throws -> [String] {
        var encoders: [String] = []
        var iter: UnsafeMutableRawPointer? = nil
        while let codec = av_codec_iterate(&iter) {
            if av_codec_is_encoder(codec) != 0 {
                if let name = codec.pointee.name {
                    encoders.append(String(cString: name))
                }
            }
        }
        return encoders
    }

    func getSupportedDecoders() throws -> [String] {
        var decoders: [String] = []
        var iter: UnsafeMutableRawPointer? = nil
        while let codec = av_codec_iterate(&iter) {
            if av_codec_is_decoder(codec) != 0 {
                if let name = codec.pointee.name {
                    decoders.append(String(cString: name))
                }
            }
        }
        return decoders
    }
}

private class CallbackContext {
    let sessionId: String
    let onProgress: (_ sessionId: String, _ progress: NativeProgress) -> Void
    let onLog: (_ sessionId: String, _ log: NativeLogEntry) -> Void

    init(
        sessionId: String,
        onProgress: @escaping (_ sessionId: String, _ progress: NativeProgress) -> Void,
        onLog: @escaping (_ sessionId: String, _ log: NativeLogEntry) -> Void
    ) {
        self.sessionId = sessionId
        self.onProgress = onProgress
        self.onLog = onLog
    }
}

File: packages/core/android/src/main/java/com/margelo/nitro/ffmpeg/HybridFFmpeg.kt
package com.margelo.nitro.ffmpeg

import com.margelo.nitro.core.Promise
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

class HybridFFmpeg : HybridFFmpegSpec() {
    private val executor = Executors.newFixedThreadPool(4)
    private val sessions = ConcurrentHashMap<String, SessionState>()

    override var onProgress: (sessionId: String, progress: NativeProgress) -> Unit = { _, _ -> }
    override var onLog: (sessionId: String, log: NativeLogEntry) -> Unit = { _, _ -> }
    override var onComplete: (sessionId: String, result: NativeSessionResult) -> Unit = { _, _ -> }

    companion object {
        init {
            System.loadLibrary("NitroFFmpeg")
        }
    }

    override fun execute(sessionId: String, args: Array<String>, logLevel: Double) {
        val state = SessionState(sessionId, args)
        sessions[sessionId] = state

        val capturedOnProgress = onProgress
        val capturedOnLog = onLog
        val capturedOnComplete = onComplete

        executor.submit {
            val startTime = System.currentTimeMillis()

            val returnCode = nativeExecute(
                sessionId,
                args,
                logLevel.toInt(),
                { sid, frame, fps, bitrate, totalSize, timeMs, speed ->
                    capturedOnProgress(sid, NativeProgress(
                        frame = frame,
                        fps = fps,
                        bitrate = bitrate,
                        totalSize = totalSize,
                        timeMs = timeMs,
                        speed = speed
                    ))
                },
                { sid, level, message ->
                    capturedOnLog(sid, NativeLogEntry(
                        level = level,
                        message = message
                    ))
                }
            )

            val duration = (System.currentTimeMillis() - startTime).toDouble()
            sessions.remove(sessionId)

            val failureMessage = if (returnCode != 0 && returnCode != 255) {
                "FFmpeg exited with code $returnCode"
            } else if (returnCode == 255) {
                "Session was cancelled"
            } else {
                null
            }

            capturedOnComplete(sessionId, NativeSessionResult(
                sessionId = sessionId,
                returnCode = returnCode.toDouble(),
                duration = duration,
                logs = "",
                command = args,
                failureMessage = failureMessage
            ))
        }
    }

    override fun cancel(sessionId: String) {
        sessions[sessionId]?.cancelled?.set(true)
        nativeCancel(sessionId)
    }

    override fun cancelAll() {
        sessions.values.forEach { it.cancelled.set(true) }
        nativeCancelAll()
    }

    override fun getActiveSessions(): Array<String> {
        return sessions.keys.toTypedArray()
    }

    override fun probe(path: String): Promise<String> {
        return Promise.async {
            nativeProbe(path)
        }
    }

    override fun getFFmpegVersion(): String {
        return nativeGetVersion()
    }

    override fun getSupportedEncoders(): Array<String> {
        return nativeGetEncoders()
    }

    override fun getSupportedDecoders(): Array<String> {
        return nativeGetDecoders()
    }

    // JNI methods
    private external fun nativeExecute(
        sessionId: String,
        args: Array<String>,
        logLevel: Int,
        progressCallback: (String, Double, Double, Double, Double, Double, Double) -> Unit,
        logCallback: (String, Double, String) -> Unit
    ): Int

    private external fun nativeCancel(sessionId: String)
    private external fun nativeCancelAll()
    private external fun nativeProbe(path: String): String
    private external fun nativeGetVersion(): String
    private external fun nativeGetEncoders(): Array<String>
    private external fun nativeGetDecoders(): Array<String>

    private data class SessionState(
        val sessionId: String,
        val args: Array<String>,
        val cancelled: AtomicBoolean = AtomicBoolean(false)
    )
}
